public/app/plugins/panel/datagrid/utils.ts

```diff
@@ -0,0 +1,304 @@
+}
+
+export interface RenameColumnInputData {
+  x?: number;
+  y?: number;
+  width?: number;
+  height?: number;
+  isInputOpen: boolean;
+  inputValue?: string;
+  columnIdx?: number;
+}
+
+interface CellRange {
+  x: number;
+  y: number;
+  width: number;
+  height: number;
+}
+
+export const getTextWidth = (text: string, isHeader = false): number => {
+  const context = TEXT_CANVAS.getContext('2d');
+  context!.font = isHeader ? HEADER_FONT_FAMILY : CELL_FONT_FAMILY;
+  const metrics = context!.measureText(text);
+  return metrics.width;
+};
+
+export const getCellWidth = (field: Field): number => {
+  //If header is longer than cell text, get header width that will always fully show the header text
+  //otherwise get the longest cell text width if it's shorter than the max column width, or the max column width
+  return Math.max(
+    getTextWidth(field.name, true) + ICON_AND_MENU_WIDTH, //header text
+    Math.min(
+      MAX_COLUMN_WIDTH,
+      field.values.toArray().reduce((acc: number, val: string | number) => {
+        const textWidth = getTextWidth(val?.toString() ?? '');
+
+        if (textWidth > acc) {
+          return textWidth;
+        }
+
+        return acc;
+      }, 0) + CELL_PADDING //cell text
+    )
+  );
+};
+
+export const deleteRows = (gridData: DataFrame, rows: number[], hardDelete = false): DataFrame => {
+  for (const field of gridData.fields) {
+    const valuesArray = field.values.toArray();
+
+    //delete from the end of the array to avoid index shifting
+    for (let i = rows.length - 1; i >= 0; i--) {
+      if (hardDelete) {
+        valuesArray.splice(rows[i], 1);
+      } else {
+        valuesArray.splice(rows[i], 1, null);
+      }
+    }
+
+    field.values = new ArrayVector(valuesArray);
+  }
+
+  return new MutableDataFrame(gridData);
+};
+
+export const clearCellsFromRangeSelection = (gridData: DataFrame, range: CellRange): DataFrame => {
+  const colFrom: number = range.x;
+  const rowFrom: number = range.y;
+  const colTo: number = range.x + range.width - 1;
+
+  for (let i = colFrom; i <= colTo; i++) {
+    const field = gridData.fields[i];
+
+    const valuesArray = field.values.toArray();
+    valuesArray.splice(rowFrom, range.height, ...new Array(range.height).fill(null));
+    field.values = new ArrayVector(valuesArray);
+  }
+
+  return new MutableDataFrame(gridData);
+};
+
+export const publishSnapshot = (data: DataFrame, panelID: number): void => {
+  if (!isDatagridEditEnabled()) {
+    return;
+  }
+
+  const snapshot: DataFrameJSON[] = [dataFrameToJSON(data)];
+  const dashboard = getDashboardSrv().getCurrent();
+  const panelModel = dashboard?.getPanelById(panelID);
+
+  const query: GrafanaQuery = {
+    refId: 'A',
+    queryType: GrafanaQueryType.Snapshot,
+    snapshot,
+    datasource: GRAFANA_DS,
+  };
+
+  panelModel!.updateQueries({
+    dataSource: GRAFANA_DS,
+    queries: [query],
+  });
+
+  panelModel!.refresh();
+
+  dashboard?.events.publish(PanelQueriesChangedEvent);
```

Comment:
@mdvictor why are you publishing this event here? It's a panel event (published by PanelModel.updateQueries)

Reply:
That should not be there. will remove